File: ProjectEuler/Problems/Problem021.cs
using System.Collections.Generic;
using System.Linq;

namespace ProjectEuler.Problems
{
    public class Problem021 : Problem<long>
    {
        public Problem021()
        {
            cache = new Dictionary<long, long>();
        }

        protected override long CalculateAnswer()
        {
            var answer = 0L;
            var n = 10000L;

            for (var i = 2L; i < n; ++i)
            {
                var sum1 = SumOfProperDivisors(i);
                var sum2 = SumOfProperDivisors(sum1);
                if (i == sum2 && i != sum1)
                {
                    answer += i;
                    WriteLineDetail("Amicable number: " + i + " (" + sum1 + "); Sum = " + answer);
                }
            }

            return answer;
        }

        private long SumOfProperDivisors(long n)
        {
            if (cache.TryGetValue(n, out long value))
            {
                return value;
            }
            value = MathHelpers.Divisors(n, true).Sum();
            cache[n] = value;
            return value;
        }

        private readonly Dictionary<long, long> cache;
    }
}

File: ProjectEulerPython/problems/problem_021.py
import math_helpers
from problems.problem import Problem

class Problem021(Problem):

  def __init__(self):
    super().__init__()
    self._cache = {}

  def calculate_answer(self) -> int:
    answer = 0
    n = 10000
    primes = math_helpers.primes_below(n)

    for i in range(2, n):
      sum_1 = self.sum_proper_divisors(i, primes)
      sum_2 = self.sum_proper_divisors(sum_1, primes)
      if i == sum_2 and i != sum_1:
        answer += i
        self.print_detail(f"amicable number: {i} ({sum_1}); sum = {answer}")

    return answer

  def sum_proper_divisors(self, n: int, primes: [int]) -> int:
    if n in self._cache:
      return self._cache[n]

    value = sum(math_helpers.divisors(n, True, primes))
    self._cache[n] = value

    return value